_posts/2022-03-16-async-sink-base.md

```diff
@@ -0,0 +1,169 @@
+
+```java
+public abstract class AsyncSinkWriter<InputT, RequestEntryT extends Serializable>
+        implements StatefulSink.StatefulSinkWriter<InputT, BufferedRequestState<RequestEntryT>> {
+    // ...
+    public AsyncSinkWriter(
+            ElementConverter<InputT, RequestEntryT> elementConverter,
+            Sink.InitContext context,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long maxBatchSizeInBytes,
+            long maxTimeInBufferMS,
+            long maxRecordSizeInBytes) { /* ... */ }
+    // ...
+}
+```
+
+By default, the method `snapshotState` returns all the elements in the buffer to be saved down for snapshots. Any elements that have been removed from the buffer for writing and are in flight have already been completed due to the sink operator calling `flush(true)` on the sink writer.
+You may want to save down additional state from the concrete sink and you can achieve this by overriding `snapshotState`, and restoring in the constructor.
+
+```java
+class MySinkWriter<InputT> extends AsyncSinkWriter<InputT, RequestEntryT> {
+
+    MySinkWriter(
+          // ... 
+          Collection<BufferedRequestState<Record>> initialStates) {
+        super(
+            // ...
+            initialStates);
+        // restore concrete sink state from initialStates
+    }
+    
+    @Override
+    public List<BufferedRequestState<RequestEntryT>> snapshotState(long checkpointId) {
+        super.snapshotState(checkpointId);
+        // ...
+    }
+
+}
+```
+
+## Sink Interface
+
+[AsyncSinkBase](https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBase.java)
+
+```java
+class MySink<InputT> extends AsyncSinkBase<InputT, RequestEntryT> {
+    // ...
+    @Override
+    public StatefulSinkWriter<InputT, BufferedRequestState<RequestEntryT>> createWriter(InitContext context) {
+        return new MySinkWriter(context);
+    }
+    // ...
+}
+```
+Sink implementers extending this will need to return their own extension of the `AsyncSinkWriter` from `createWriter()` inside their own implementation of `AsyncSinkBase`.
+
+At the time of writing, the [Kinesis Data Streams sink](https://github.com/apache/flink/tree/master/flink-connectors/flink-connector-aws-kinesis-streams) and [Kinesis Data Firehose sink](https://github.com/apache/flink/tree/master/flink-connectors/flink-connector-aws-kinesis-firehose) are using this base sink. 
+
+# Metrics
+
+There are three metrics that automatically exist when you implement sinks (and, thus, should not be implemented by yourself).
+
+* CurrentSendTime Gauge - returns the amount of time in milliseconds it took for the most recent request to write records to return, whether successful or not.  
```

Comment:
"return" is throwing me off a bit. Is it actually the time that it took for submitRequestEntries() to _return_, or for `requestResult#accept` to be called?